approaches/abstract_approach.py

```diff
@@ -138,3 +167,21 @@ def get_duplicated_ids(self, query_num: int, topn: int) -> np.ndarray:
     @abstractmethod
     def update_history(self, query_num: int):
         raise NotImplementedError()
+
+    def __get_relevant_reports_num(self) -> Dict[str, int]:
+        result: Dict[str, int] = dict()
+
+        query_ids = self.test.id.tolist()
+        train_disc_ids = self.train.disc_id.tolist()
+        test_disc_ids = self.test.disc_id.tolist()
```

Comment:
What is the structure of these dataframes? It looks like the iteration might be made more efficiently

Reply:
The structure is something like this:

|  id  | ... |   summary   |  ...  |  description  |  ...  | disc_id |
|:----:|-----|:-----------:|:-----:|:-------------:|:-----:|:-------:|
| ...  | ... |     ...     |  ...  |      ...      |  ...  |   ...   |
|  33  | ... | bug example |  ...  | this is a bug |  ...  |   11    |

I tried to optimise this method in this way:

```python
  def __get_relevant_reports_num(self) -> Counter:
      result: Counter = Counter()
      train_disc_ids = self.train.disc_id.tolist()
      def collect_relevant_reports(query):
          for train_disc_id in train_disc_ids:
              result[query.id] += (query.disc_id == train_disc_id)
          train_disc_ids.append(query.disc_id)

      self.test.apply(collect_relevant_reports, axis=1)

      return result
```

but it starts to work much slower. So I left the old implementation.

Reply:
Ok, I did some optimizations, now it looks like this:
```python
    def __get_relevant_reports_num(self) -> Counter:
        result: Counter = Counter()

        test_disc_ids = self.test.disc_id.tolist()
        count_train: Counter = Counter(self.train.disc_id)

        for i, query_id in enumerate(self.test.id):
            result[query_id] += count_train[test_disc_ids[i]]
            count_train[test_disc_ids[i]] += 1

        return result
```